.github/workflows/continuous-integration.yml

```
@@ -243,7 +279,13 @@ jobs:
   test_e2e:
     runs-on: ubuntu-22.04
     name: Test E2E
-    needs: [ compile_frontend, compile_backend ]
+    needs: [detect_changes, test_frontend, test_backend_17, test_backend_21]
```

This is redundant due to transitive closure, right?

```suggestion
    needs: [test_frontend, test_backend_17, test_backend_21]
```

---

This is needed because the job’s condition depends on the detect_changes output. Without using it here, the workflow would fail with an error.

---

Yeah but as the other elements of `needs: [..]` all already depend on `detect_changes` it's redundant here - hence why we can make use of the transitive closure.

We already use the closure for `test_frontend` and `test_backend_*`, don't we?
These don't directly depend on `detect_changes` but their other _needs_ - the compile-jobs - do.

---

I don't have a strong opinion about this. But at least we should be consistent. In `release_artifacts`, `deploy_to_azure` and `upload_to_sonar` the `detect_changes` is not part of the `needs` definition.

---

the problem is that we are using the output of `detect_changes` in `Test E2E` so we need it in the `needs` if not we cannot access `needs.detect_changes.outputs.code_changed` or any other output

---

Fair point!

---

Now i see what you mean in `deploy_to_azure` , `deploy_to_azure` and `upload_to_sonar` we are using output of `detect_changes` and it is not in needs i will verify if it is really optional
my answers was folowing the documentation [needs-context](https://docs.github.com/en/actions/reference/workflows-and-actions/contexts#needs-context) they say "Note that this doesn't include implicitly dependent jobs"